app/guard/guard.go

```diff
@@ -0,0 +1,315 @@
+
+import (
+	"os"
+
+	"errors"
+
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+
+	"github.com/dedis/cothority/app/lib/config"
+	"github.com/dedis/cothority/log"
+	"github.com/dedis/cothority/sda"
+	"github.com/dedis/cothority/services/guard"
+	"github.com/dedis/crypto/abstract"
+
+	"io/ioutil"
+
+	s "github.com/SSSaaS/sssa-golang"
+	"gopkg.in/codegangsta/cli.v1"
+
+	"bytes"
+
+	"github.com/dedis/cothority/network"
+)
+
+// User is a representation of the Users data in the code, and is used to store
+// all relevant information.
+type User struct {
+	// Name or UserID
+	Name []byte
+	// Salt used for the password-hash
+	Salt []byte
+	// Xored keys with hash
+	Keys [][]byte
+	// Data AEAD-encrypted with key
+	Data []byte
+	Iv   []byte
+}
+
+// Database is a structure that stores Cothority(the list of guard servers), and
+// a list of all users within the database.
+type Database struct {
+	Cothority *sda.Roster
+	Users     []User
+}
+
+var db *Database
+
+func main() {
+	network.RegisterMessageType(&Database{})
+
+	app := cli.NewApp()
+	app.Name = "Guard"
+	app.Usage = "Get and print status of all servers of a file."
+
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:  "group, gd",
+			Value: "group.toml",
+			Usage: "Cothority group definition in `FILE.toml`",
+		},
+		cli.IntFlag{
+			Name:  "debug, d",
+			Value: 0,
+			Usage: "debug-level: `integer`: 1 for terse, 5 for maximal",
+		},
+	}
+	app.Commands = []cli.Command{
+		{
+			Name:    "setpass",
+			Aliases: []string{"s"},
+			Usage:   "Setup the configuration for the server (interactive)",
+			Action:  setpass,
+		},
+		{
+			Name:      "setup",
+			Aliases:   []string{"su"},
+			Usage:     "Saves the cothority group-toml to the configuration",
+			ArgsUsage: "Give group definition",
+			Action:    setup,
+		},
+		{
+			Name:    "recover",
+			Aliases: []string{"r"},
+			Usage:   "Gets the password back from the guards",
+			Action:  get,
+		},
+	}
+	app.Before = func(c *cli.Context) error {
+		b, err := ioutil.ReadFile("config.bin")
+		if os.IsNotExist(err) {
+			return nil
+		}
+		log.ErrFatal(err, "The config.bin file threw an error")
+		_, msg, err := network.UnmarshalRegistered(b)
+		log.ErrFatal(err, "UnmarshalRegistered messeed up")
+		var ok bool
+		db, ok = msg.(*Database)
+		if !ok {
+			log.Fatal("We are not OK")
+		}
+		return nil
+	}
+	app.Run(os.Args)
+}
+
+// readGroup takes a toml file name and reads the file, returning the entities within.
+func readGroup(tomlFileName string) (*sda.Roster, error) {
+	log.Lvl2("Reading From File")
+	f, err := os.Open(tomlFileName)
+	if err != nil {
+		return nil, err
+	}
+	el, err := config.ReadGroupToml(f)
+	if err != nil {
+		return nil, err
+	}
+	if len(el.List) <= 0 {
+		return nil, errors.New("Empty or invalid group file:" +
+			tomlFileName)
+	}
+	log.Lvl3(el)
+	return el, err
+}
+
+func set(c *cli.Context, uid []byte, epoch []byte, password string, userdata []byte) {
+	mastersalt := make([]byte, 12)
+	rand.Read(mastersalt)
```

Comment:
In very rare and ugly cases (almost impossible) rand.Read can return an error (not enough entropy). You would definitely want to catch this cases (and `panic` to be safe) .